### PR TITLE
Handle failure if branch not pushed

### LINKED ISSUE #161: Github raises an exception if review is not well formed
When creating a review (`zazu dev review`) should ensure that the feature branch is uploaded.

### PR DESCRIPTION
**Pull request by aramnhammer**
zazu dev review should check for dirty working tree and push if the branch has not been pushed

Fixes #161

### CODE REVIEW

File: zazu/dev/commands.py [hunk 1]
```diff
@@ -309,6 +309,13 @@ def review(config, base, head):
         pr = zazu.util.pick(existing_reviews, 'Multiple reviews found, pick one')
     else:
         descriptor = make_issue_descriptor(head)
+        dirty = config.repo.git.status(['--porcelain'])
+        if dirty:
+            raise click.ClickException('working tree is not clean, stash or remove changes before review')    
+        try:
+            config.repo.git.push('--set-upstream','origin', head)
```
Comment (stopthatcow) [inline zazu/dev/commands.py]: click.echo('Pushing to origin...')
            config.repo.git.push('--set-upstream', 'origin', head)

Comment (aramnhammer) [reply zazu/dev/commands.py]: 3950ac2

File: tests/test_dev.py [hunk 2]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
+    mocker.patch('zazu.config.Config.code_reviewer', return_value=mocked_reviewer)
+    mocker.patch('zazu.util.prompt', side_effect=['title', 'summary'])
+    with zazu.util.cd(git_repo.working_tree_dir):
+        runner = click.testing.CliRunner()
+        result = runner.invoke(zazu.cli.cli, ['dev', 'review'])
+        assert result.exception
+        assert result.exit_code == 1
+
+
+def test_review_dirty_working_tree(mocker, git_repo_with_local_origin, tmp_dir):
```
Comment (stopthatcow) [inline tests/test_dev.py:275]: I don't think we need this?

File: tests/test_dev.py [hunk 3]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
+    mocked_tracker.issue = mocker.Mock(side_effect=zazu.issue_tracker.IssueTrackerError)
+    mocked_reviewer = mocker.Mock()
+    mocked_reviewer.review = mocker.Mock(return_value=[])
+    mocked_reviewer.create_review = mocker.Mock()
+    mocker.patch('zazu.config.Config.issue_tracker', return_value=mocked_tracker)
+    mocker.patch('zazu.config.Config.code_reviewer', return_value=mocked_reviewer)
+    mocker.patch('zazu.util.prompt', side_effect=['title', 'summary'])
+    pathlib.Path(tmp_dir).joinpath('un-tracked_file.txt').touch()
+    with zazu.util.cd(git_repo_with_local_origin.working_tree_dir):
+        git_repo_with_local_origin
```
Comment (stopthatcow) [inline tests/test_dev.py]: Un-needed?

File: tests/test_dev.py [hunk 4]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
+def test_review_push_fails(mocker, git_repo):
+    mocker.patch('webbrowser.open_new')
+    mocked_tracker = mocker.Mock()
+    mocked_tracker.issue = mocker.Mock(side_effect=zazu.issue_tracker.IssueTrackerError)
+    mocked_reviewer = mocker.Mock()
+    mocked_reviewer.review = mocker.Mock(return_value=[])
+    mocked_reviewer.create_review = mocker.Mock()
+    mocker.patch('zazu.config.Config.issue_tracker', return_value=mocked_tracker)
+    mocker.patch('zazu.config.Config.code_reviewer', return_value=mocked_reviewer)
+    mocker.patch('zazu.util.prompt', side_effect=['title', 'summary'])
```
Comment (stopthatcow) [inline tests/test_dev.py]: Not needed as it will never be reached.

File: tests/test_dev.py [hunk 5]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
 
 
+def test_review_push_fails(mocker, git_repo):
+    mocker.patch('webbrowser.open_new')
+    mocked_tracker = mocker.Mock()
+    mocked_tracker.issue = mocker.Mock(side_effect=zazu.issue_tracker.IssueTrackerError)
+    mocked_reviewer = mocker.Mock()
+    mocked_reviewer.review = mocker.Mock(return_value=[])
+    mocked_reviewer.create_review = mocker.Mock()
+    mocker.patch('zazu.config.Config.issue_tracker', return_value=mocked_tracker)
```
Comment (stopthatcow) [inline tests/test_dev.py]: Not needed as it will never be reached.

File: zazu/dev/commands.py (hunk 1 above)
Comment (stopthatcow) [reply zazu/dev/commands.py]: Still missing space between '--set-upstream','origin' try running `zazu style` in the zazu repo to fix and if that works, run `zazu repo init` in your zazu repo to install git hooks.

File: tests/test_dev.py [hunk 6]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
+    mocker.patch('zazu.config.Config.code_reviewer', return_value=mocked_reviewer)
+    mocker.patch('zazu.util.prompt', side_effect=['title', 'summary'])
+    pathlib.Path(tmp_dir).joinpath('un-tracked_file.txt').touch()
```
Comment (stopthatcow) [inline tests/test_dev.py]: simpler to do after `zazu.util.cd` and remove the import of pathlib:
`with open('un-tracked_file.txt', 'w') as f: pass`

File: tests/test_dev.py [hunk 7]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
+        runner = click.testing.CliRunner()
+        result = runner.invoke(zazu.cli.cli, ['dev', 'review'])
+        assert result.exception
+        assert result.exit_code == 1
+
+
+def test_review_dirty_working_tree(mocker, git_repo_with_local_origin, tmp_dir):
+    mocker.patch('webbrowser.open_new')
+    mocked_tracker = mocker.Mock()
+    mocked_tracker.issue = mocker.Mock(side_effect=zazu.issue_tracker.IssueTrackerError)
```
Comment (stopthatcow) [inline tests/test_dev.py]: Not needed.

File: tests/test_dev.py [hunk 8]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
+    mocked_reviewer = mocker.Mock()
+    mocked_reviewer.review = mocker.Mock(return_value=[])
+    mocked_reviewer.create_review = mocker.Mock()
+    mocker.patch('zazu.config.Config.issue_tracker', return_value=mocked_tracker)
```
Comment (stopthatcow) [inline tests/test_dev.py]: Not needed?

File: tests/test_dev.py [hunk 9]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
+    mocker.patch('zazu.config.Config.code_reviewer', return_value=mocked_reviewer)
+    mocker.patch('zazu.util.prompt', side_effect=['title', 'summary'])
```
Comment (stopthatcow) [inline tests/test_dev.py]: Not needed

File: tests/test_dev.py [hunk 10]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
-    with zazu.util.cd(git_repo.working_tree_dir):
+    with zazu.util.cd(git_repo_with_local_origin.working_tree_dir):
         runner = click.testing.CliRunner()
         result = runner.invoke(zazu.cli.cli, ['dev', 'review'])
         assert not result.exception
         assert result.exit_code == 0
 
 
+def test_review_push_fails(mocker, git_repo):
+    mocker.patch('webbrowser.open_new')
```
Comment (stopthatcow) [inline tests/test_dev.py]: Not needed

File: tests/test_dev.py [hunk 11]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
+    mocked_tracker = mocker.Mock()
```
Comment (stopthatcow) [inline tests/test_dev.py]: Not needed as it will never be reached.

File: tests/test_dev.py [hunk 12]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
+    mocked_tracker.issue = mocker.Mock(side_effect=zazu.issue_tracker.IssueTrackerError)
```
Comment (stopthatcow) [inline tests/test_dev.py]: Not needed as it will never be reached.

File: tests/test_dev.py [hunk 13]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
+    mocked_reviewer = mocker.Mock()
+    mocked_reviewer.review = mocker.Mock(return_value=[])
+    mocked_reviewer.create_review = mocker.Mock()
```
Comment (stopthatcow) [inline tests/test_dev.py:265]: Should we assert that this is not called?

File: tests/test_dev.py [hunk 14]
```diff
@@ -252,13 +253,49 @@ def test_review(mocker, git_repo):
...
+        assert result.exit_code == 1
+
+
+def test_review_dirty_working_tree(mocker, git_repo_with_local_origin, tmp_dir):
+    mocker.patch('webbrowser.open_new')
+    mocked_tracker = mocker.Mock()
+    mocked_tracker.issue = mocker.Mock(side_effect=zazu.issue_tracker.IssueTrackerError)
+    mocked_reviewer = mocker.Mock()
+    mocked_reviewer.review = mocker.Mock(return_value=[])
+    mocked_reviewer.create_review = mocker.Mock()
```
Comment (stopthatcow) [inline tests/test_dev.py:278]: Should we assert that this is not called?

**Review by stopthatcow — CHANGES_REQUESTED** [review]

File: tests/test_dev.py (hunk 2 above)
Comment (aramnhammer) [reply tests/test_dev.py:275]: What covers that code path? This test covers the exception that should get thrown when the working tree is dirty.

File: tests/test_dev.py (hunk 10 above)
Comment (aramnhammer) [reply tests/test_dev.py]: 631e1d1

File: zazu/dev/commands.py (hunk 1 above)
Comment (aramnhammer) [reply zazu/dev/commands.py]: 23af5dd

File: tests/test_dev.py (hunk 2 above)
Comment (stopthatcow) [reply tests/test_dev.py:275]: Wrong line.

File: tests/test_dev.py (hunk 14 above)
Comment (aramnhammer) [reply tests/test_dev.py:278]: ddc0b68

Comment (stopthatcow) [reply tests/test_dev.py:278]: Better to use `mocked_reviewer.create_review.assert_not_called()` here and in the test above

File: tests/test_dev.py [hunk 15]
```diff
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 import click
 import click.testing
+import git
```
Comment (stopthatcow) [inline tests/test_dev.py]: Unused?

File: .gitignore [hunk 16]
```diff
@@ -91,3 +91,6 @@ ENV/
 # Rope project settings
 .ropeproject
 version.txt
+
+# vscode project settings
```
Comment (stopthatcow) [inline .gitignore]: `# Visual Studio Code project settings`

File: tests/test_dev.py (hunk 15 above)
Comment (aramnhammer) [reply tests/test_dev.py]: man i am so use to pycharm... maybe we can make a pre-commit hook to clean up unused imports?

307863c

File: .gitignore (hunk 16 above)
Comment (aramnhammer) [reply .gitignore]: 576296a

**Review by stopthatcow — APPROVED** [review]

**Pull request merged**